ICE: `ConstArgHasWrongType failed but we don't know how to compute type`

<!--
[31mICE[0m: Rustc ./a.rs '' 'error: internal compiler error: compiler/rustc_trait_selection/src/solve/fulfill.rs:267:22: ConstArgHasWrongType failed but we don't know how to compute type', 'error: internal compiler error: compiler/rustc_trait_selection/src/solve/fulfill.rs:267:22: ConstArgHasWrongType failed but we don't know how to compute type'
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait If<const generic_const_exprs: bool> {}
impl If<true> for () {}

trait Foobar {}

impl<const N: u8> Foobar<N> for () where (): If<N> {}

impl Foobar<N> for () {}

````

original:
````rust
#![feature(generic_const_exprs)]

trait If<const generic_const_exprs: bool> {}
impl If<true> for () {}

trait IsZero<const N: usize> {}

impl<const N: usize> IsZero<N> for () where (If): If<{ N == 0 }> {}

trait Foobar<const N: u8> {}

impl<const N: u8> Foobar<N> for () where (): If<N> {}

impl<const N: u8> Foobar<N> for () {}

pub fn main() {}

````

Version information
````
rustc 1.86.0-nightly (7e4077d06 2025-01-11)
binary: rustc
commit-hash: 7e4077d06fc073442c567d58885b47ed2c5121b8
commit-date: 2025-01-11
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.6

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/7e4077d06fc073442c567d58885b47ed2c5121b8/compiler/rustc_trait_selection/src/solve/fulfill.rs#L261-L273

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0412]: cannot find type `N` in this scope
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:8:13
  |
8 | impl Foobar<N> for () {}
  |             ^ not found in this scope
  |
help: you might be missing a type parameter
  |
8 | impl<N> Foobar<N> for () {}
  |     +++

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:8:25
  |
8 | impl Foobar<N> for () {}
  |                         ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs`

error[E0107]: trait takes 0 generic arguments but 1 generic argument was supplied
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:6:19
  |
6 | impl<const N: u8> Foobar<N> for () where (): If<N> {}
  |                   ^^^^^^--- help: remove the unnecessary generics
  |                   |
  |                   expected 0 generic arguments
  |
note: trait defined here, with 0 generic parameters
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:4:7
  |
4 | trait Foobar {}
  |       ^^^^^^

error[E0207]: the const parameter `N` is not constrained by the impl trait, self type, or predicates
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:6:6
  |
6 | impl<const N: u8> Foobar<N> for () where (): If<N> {}
  |      ^^^^^^^^^^^ unconstrained const parameter
  |
  = note: expressions using a const parameter must map each value to a distinct output value
  = note: proving the result of expressions other than the parameter are unique is not supported

error[E0107]: trait takes 0 generic arguments but 1 generic argument was supplied
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:8:6
  |
8 | impl Foobar<N> for () {}
  |      ^^^^^^--- help: remove the unnecessary generics
  |      |
  |      expected 0 generic arguments
  |
note: trait defined here, with 0 generic parameters
 --> /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:4:7
  |
4 | trait Foobar {}
  |       ^^^^^^

error: internal compiler error: compiler/rustc_trait_selection/src/solve/fulfill.rs:267:22: ConstArgHasWrongType failed but we don't know how to compute type


thread 'rustc' panicked at compiler/rustc_trait_selection/src/solve/fulfill.rs:267:22:
Box<dyn Any>
stack backtrace:
   0:     0x75e2d90f5b9a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h6ad413a9c5d92132
   1:     0x75e2d9812d66 - core::fmt::write::hd9d1731ab47c0ca7
   2:     0x75e2da777e91 - std::io::Write::write_fmt::h7f0b3d462e354385
   3:     0x75e2d90f59f2 - std::sys::backtrace::BacktraceLock::print::h2b918afb2877943f
   4:     0x75e2d90f7f97 - std::panicking::default_hook::{{closure}}::h6129e8f978485814
   5:     0x75e2d90f7d80 - std::panicking::default_hook::h03e9c5e42fe92791
   6:     0x75e2d825df38 - std[33acbbc1924c4cd6]::panicking::update_hook::<alloc[893ed522702f9859]::boxed::Box<rustc_driver_impl[f9fe3354fe2fd435]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x75e2d90f87e3 - std::panicking::rust_panic_with_hook::h9ba2b54837e008ba
   8:     0x75e2d8296d81 - std[33acbbc1924c4cd6]::panicking::begin_panic::<rustc_errors[2805a1dbbffdc483]::ExplicitBug>::{closure#0}
   9:     0x75e2d828bc86 - std[33acbbc1924c4cd6]::sys::backtrace::__rust_end_short_backtrace::<std[33acbbc1924c4cd6]::panicking::begin_panic<rustc_errors[2805a1dbbffdc483]::ExplicitBug>::{closure#0}, !>
  10:     0x75e2d8288799 - std[33acbbc1924c4cd6]::panicking::begin_panic::<rustc_errors[2805a1dbbffdc483]::ExplicitBug>
  11:     0x75e2d82a0cc1 - <rustc_errors[2805a1dbbffdc483]::diagnostic::BugAbort as rustc_errors[2805a1dbbffdc483]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x75e2d87f780c - <rustc_errors[2805a1dbbffdc483]::DiagCtxtHandle>::span_bug::<rustc_span[475c1a9bb47d94c0]::span_encoding::Span, alloc[893ed522702f9859]::string::String>
  13:     0x75e2d887ceb7 - rustc_middle[ff83cc293d1a3dd6]::util::bug::opt_span_bug_fmt::<rustc_span[475c1a9bb47d94c0]::span_encoding::Span>::{closure#0}
  14:     0x75e2d8861eba - rustc_middle[ff83cc293d1a3dd6]::ty::context::tls::with_opt::<rustc_middle[ff83cc293d1a3dd6]::util::bug::opt_span_bug_fmt<rustc_span[475c1a9bb47d94c0]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x75e2d8861d4b - rustc_middle[ff83cc293d1a3dd6]::ty::context::tls::with_context_opt::<rustc_middle[ff83cc293d1a3dd6]::ty::context::tls::with_opt<rustc_middle[ff83cc293d1a3dd6]::util::bug::opt_span_bug_fmt<rustc_span[475c1a9bb47d94c0]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x75e2d72a50f7 - rustc_middle[ff83cc293d1a3dd6]::util::bug::span_bug_fmt::<rustc_span[475c1a9bb47d94c0]::span_encoding::Span>
  17:     0x75e2da7e08e2 - <rustc_trait_selection[a9962fce8e4fbdac]::traits::FulfillmentError as rustc_infer[4d6832c29a414c7]::traits::engine::FromSolverError<rustc_trait_selection[a9962fce8e4fbdac]::solve::fulfill::NextSolverError>>::from_solver_error
  18:     0x75e2da5478a3 - <rustc_trait_selection[a9962fce8e4fbdac]::solve::fulfill::FulfillmentCtxt<rustc_trait_selection[a9962fce8e4fbdac]::traits::FulfillmentError> as rustc_infer[4d6832c29a414c7]::traits::engine::TraitEngine<rustc_trait_selection[a9962fce8e4fbdac]::traits::FulfillmentError>>::select_where_possible
  19:     0x75e2da7dfdb3 - <rustc_trait_selection[a9962fce8e4fbdac]::solve::fulfill::FulfillmentCtxt<rustc_trait_selection[a9962fce8e4fbdac]::traits::FulfillmentError> as rustc_infer[4d6832c29a414c7]::traits::engine::TraitEngine<rustc_trait_selection[a9962fce8e4fbdac]::traits::FulfillmentError>>::select_all_or_error
  20:     0x75e2da542977 - rustc_trait_selection[a9962fce8e4fbdac]::traits::coherence::overlap
  21:     0x75e2d981a5bd - <rustc_middle[ff83cc293d1a3dd6]::traits::specialization_graph::Children as rustc_trait_selection[a9962fce8e4fbdac]::traits::specialize::specialization_graph::ChildrenExt>::insert
  22:     0x75e2da69565b - <rustc_middle[ff83cc293d1a3dd6]::traits::specialization_graph::Graph as rustc_trait_selection[a9962fce8e4fbdac]::traits::specialize::specialization_graph::GraphExt>::insert
  23:     0x75e2d9a3af33 - rustc_trait_selection[a9962fce8e4fbdac]::traits::specialize::specialization_graph_provider
  24:     0x75e2d9a3a955 - rustc_query_impl[85545818dc61c388]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[85545818dc61c388]::query_impl::specialization_graph_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 8usize]>>
  25:     0x75e2d9b119df - rustc_query_system[861d73d81f5ebaeb]::query::plumbing::try_execute_query::<rustc_query_impl[85545818dc61c388]::DynamicConfig<rustc_query_system[861d73d81f5ebaeb]::query::caches::DefIdCache<rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[85545818dc61c388]::plumbing::QueryCtxt, false>
  26:     0x75e2da45cb81 - rustc_query_impl[85545818dc61c388]::query_impl::specialization_graph_of::get_query_non_incr::__rust_end_short_backtrace
  27:     0x75e2d9e55165 - rustc_hir_analysis[b4134b0e1e97b436]::coherence::coherent_trait
  28:     0x75e2d9e54ecf - rustc_query_impl[85545818dc61c388]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[85545818dc61c388]::query_impl::coherent_trait::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>
  29:     0x75e2d9bf3830 - rustc_query_system[861d73d81f5ebaeb]::query::plumbing::try_execute_query::<rustc_query_impl[85545818dc61c388]::DynamicConfig<rustc_query_system[861d73d81f5ebaeb]::query::caches::DefIdCache<rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[85545818dc61c388]::plumbing::QueryCtxt, false>
  30:     0x75e2d9bf2862 - rustc_query_impl[85545818dc61c388]::query_impl::coherent_trait::get_query_non_incr::__rust_end_short_backtrace
  31:     0x75e2d9e95e8c - rustc_middle[ff83cc293d1a3dd6]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[861d73d81f5ebaeb]::query::caches::DefIdCache<rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>, ()>
  32:     0x75e2d9ea1f2c - rustc_hir_analysis[b4134b0e1e97b436]::check::wfcheck::check_well_formed
  33:     0x75e2d9ea16a3 - rustc_query_impl[85545818dc61c388]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[85545818dc61c388]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>
  34:     0x75e2d9ea1145 - rustc_query_system[861d73d81f5ebaeb]::query::plumbing::try_execute_query::<rustc_query_impl[85545818dc61c388]::DynamicConfig<rustc_data_structures[6294f6aa86e67250]::vec_cache::VecCache<rustc_span[475c1a9bb47d94c0]::def_id::LocalDefId, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[861d73d81f5ebaeb]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[85545818dc61c388]::plumbing::QueryCtxt, false>
  35:     0x75e2d9ea0bc1 - rustc_query_impl[85545818dc61c388]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  36:     0x75e2d9e9dfec - rustc_hir_analysis[b4134b0e1e97b436]::check::wfcheck::check_mod_type_wf
  37:     0x75e2d9e9de0b - rustc_query_impl[85545818dc61c388]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[85545818dc61c388]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>
  38:     0x75e2da7b9e88 - rustc_query_system[861d73d81f5ebaeb]::query::plumbing::try_execute_query::<rustc_query_impl[85545818dc61c388]::DynamicConfig<rustc_query_system[861d73d81f5ebaeb]::query::caches::DefaultCache<rustc_span[475c1a9bb47d94c0]::def_id::LocalModDefId, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[85545818dc61c388]::plumbing::QueryCtxt, false>
  39:     0x75e2da7b9c30 - rustc_query_impl[85545818dc61c388]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  40:     0x75e2d99ff338 - rustc_hir_analysis[b4134b0e1e97b436]::check_crate
  41:     0x75e2d9bec3e8 - rustc_interface[3ef31e66eb8873de]::passes::run_required_analyses
  42:     0x75e2da77bd1e - rustc_interface[3ef31e66eb8873de]::passes::analysis
  43:     0x75e2da77bcef - rustc_query_impl[85545818dc61c388]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[85545818dc61c388]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 0usize]>>
  44:     0x75e2da7d8955 - rustc_query_system[861d73d81f5ebaeb]::query::plumbing::try_execute_query::<rustc_query_impl[85545818dc61c388]::DynamicConfig<rustc_query_system[861d73d81f5ebaeb]::query::caches::SingleCache<rustc_middle[ff83cc293d1a3dd6]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[85545818dc61c388]::plumbing::QueryCtxt, false>
  45:     0x75e2da7d868e - rustc_query_impl[85545818dc61c388]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  46:     0x75e2da82355e - rustc_interface[3ef31e66eb8873de]::passes::create_and_enter_global_ctxt::<core[6cf442861785187b]::option::Option<rustc_interface[3ef31e66eb8873de]::queries::Linker>, rustc_driver_impl[f9fe3354fe2fd435]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  47:     0x75e2da79eba4 - rustc_interface[3ef31e66eb8873de]::interface::run_compiler::<(), rustc_driver_impl[f9fe3354fe2fd435]::run_compiler::{closure#0}>::{closure#1}
  48:     0x75e2da6908d1 - std[33acbbc1924c4cd6]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[3ef31e66eb8873de]::util::run_in_thread_with_globals<rustc_interface[3ef31e66eb8873de]::util::run_in_thread_pool_with_globals<rustc_interface[3ef31e66eb8873de]::interface::run_compiler<(), rustc_driver_impl[f9fe3354fe2fd435]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  49:     0x75e2da690d84 - <<std[33acbbc1924c4cd6]::thread::Builder>::spawn_unchecked_<rustc_interface[3ef31e66eb8873de]::util::run_in_thread_with_globals<rustc_interface[3ef31e66eb8873de]::util::run_in_thread_pool_with_globals<rustc_interface[3ef31e66eb8873de]::interface::run_compiler<(), rustc_driver_impl[f9fe3354fe2fd435]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[6cf442861785187b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  50:     0x75e2da692341 - std::sys::pal::unix::thread::Thread::new::thread_start::h8bb747bec67b46a6
  51:     0x75e2d4aa339d - <unknown>
  52:     0x75e2d4b2849c - <unknown>
  53:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.86.0-nightly (7e4077d06 2025-01-11) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [specialization_graph_of] building specialization graph of trait `Foobar`
#1 [coherent_trait] coherence checking all impls of trait `Foobar`
#2 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:8:1: 8:22>` is well-formed
#3 [check_mod_type_wf] checking that types are well-formed in top-level module
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0107, E0207, E0412, E0601.
For more information about an error, try `rustc --explain E0107`.

```

</p>
</details>

<!--
query stack:
#0 [specialization_graph_of] building specialization graph of trait `Foobar`
#1 [coherent_trait] coherence checking all impls of trait `Foobar`
#2 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.Clgh7OT0H3Q5/rustc_testrunner_tmpdir_reporting.8bDVhjbOqSrL/mvce.rs:8:1: 8:22>` is well-formed
#3 [check_mod_type_wf] checking that types are well-formed in top-level module
#4 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_trait_selection/src/solve/fulfill.rs
+++ b/compiler/rustc_trait_selection/src/solve/fulfill.rs
@@ -264,9 +264,10 @@ fn fulfillment_error_for_no_solution<'tcx>(
                     infcx.tcx.type_of(uv.def).instantiate(infcx.tcx, uv.args)
                 }
                 ty::ConstKind::Param(param_ct) => param_ct.find_ty_from_env(obligation.param_env),
-                _ => span_bug!(
+                ty::ConstKind::Value(ty, _) => ty,
+                kind => span_bug!(
                     obligation.cause.span,
-                    "ConstArgHasWrongType failed but we don't know how to compute type"
+                    "ConstArgHasWrongType failed but we don't know how to compute type for {kind:?}"
                 ),
             };
             FulfillmentErrorCode::Select(SelectionError::ConstArgHasWrongType {

--- a/tests/ui/const-generics/type-mismatch-in-nested-goal.current.stderr
+++ b/tests/ui/const-generics/type-mismatch-in-nested-goal.current.stderr
@@ -1,0 +1,45 @@
+error: the constant `N` is not of type `bool`
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:50
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |                                                  ^^^^ expected `bool`, found `usize`
+   |
+note: required by a const generic parameter in `A`
+  --> $DIR/type-mismatch-in-nested-goal.rs:5:9
+   |
+LL | trait A<const B: bool> {}
+   |         ^^^^^^^^^^^^^ required by this const generic parameter in `A`
+
+error: the constant `true` is not of type `usize`
+  --> $DIR/type-mismatch-in-nested-goal.rs:13:13
+   |
+LL |     needs_a([]);
+   |     ------- ^^ expected `usize`, found `bool`
+   |     |
+   |     required by a bound introduced by this call
+   |
+note: required by a const generic parameter in `needs_a`
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:12
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |            ^^^^^^^^^^^^^^ required by this const generic parameter in `needs_a`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch-in-nested-goal.rs:13:13
+   |
+LL |     needs_a([]);
+   |     ------- ^^ expected an array with a size of true, found one with a size of 0
+   |     |
+   |     arguments to this function are incorrect
+   |
+   = note: expected array `[u8; true]`
+              found array `[_; 0]`
+note: function defined here
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:4
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |    ^^^^^^^                 ----------
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/const-generics/type-mismatch-in-nested-goal.next.stderr
+++ b/tests/ui/const-generics/type-mismatch-in-nested-goal.next.stderr
@@ -1,0 +1,45 @@
+error: the constant `N` is not of type `bool`
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:50
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |                                                  ^^^^ expected `bool`, found `usize`
+   |
+note: required by a const generic parameter in `A`
+  --> $DIR/type-mismatch-in-nested-goal.rs:5:9
+   |
+LL | trait A<const B: bool> {}
+   |         ^^^^^^^^^^^^^ required by this const generic parameter in `A`
+
+error: the constant `true` is not of type `usize`
+  --> $DIR/type-mismatch-in-nested-goal.rs:13:13
+   |
+LL |     needs_a([]);
+   |     ------- ^^ expected `usize`, found `bool`
+   |     |
+   |     required by a bound introduced by this call
+   |
+note: required by a const generic parameter in `needs_a`
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:12
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |            ^^^^^^^^^^^^^^ required by this const generic parameter in `needs_a`
+
+error[E0308]: mismatched types
+  --> $DIR/type-mismatch-in-nested-goal.rs:13:13
+   |
+LL |     needs_a([]);
+   |     ------- ^^ expected an array with a size of true, found one with a size of 0
+   |     |
+   |     arguments to this function are incorrect
+   |
+   = note: expected array `[u8; true]`
+              found array `[_; 0]`
+note: function defined here
+  --> $DIR/type-mismatch-in-nested-goal.rs:9:4
+   |
+LL | fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+   |    ^^^^^^^                 ----------
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/const-generics/type-mismatch-in-nested-goal.rs
+++ b/tests/ui/const-generics/type-mismatch-in-nested-goal.rs
@@ -1,0 +1,17 @@
+//@ revisions: current next
+//@[next] compile-flags: -Znext-solver
+//@ ignore-compare-mode-next-solver (explicit revisions)
+
+trait A<const B: bool> {}
+
+impl A<true> for () {}
+
+fn needs_a<const N: usize>(_: [u8; N]) where (): A<N> {}
+//~^ ERROR the constant `N` is not of type `bool`
+
+pub fn main() {
+    needs_a([]);
+    //~^ ERROR the constant `true` is not of type `usize`
+    //~| ERROR mismatched types
+    // FIXME(const_generics): we should hide this error as we've already errored above
+}

